lib/types.nix

```diff
@@ -358,25 +358,41 @@ rec {
     };
 
     # A submodule (like typed attribute set). See NixOS manual.
-    submodule = opts:
+    submodule = modules: submoduleWith {
+      shorthandOnlyDefinesConfig = true;
+      modules = toList modules;
+    };
+
+    submoduleWith =
+      { shorthandOnlyDefinesConfig ? false
+      , modules ? []
+      , specialArgs ? {}
+      , hideSubOptions ? false
```

Comment:
I'd like to suggest `pruneSubOptions` here to distinguish it more from 'visible = false` and `internal = true`.

Reply:
Nice, did that